### PR TITLE
Quickcheck test for vnode proxy overload recovery.

### PR DESCRIPTION
**Pull request by jonmeredith**
Requires vnode proxy fix (probably release *after* 1.4.12/2.0.6/2.1.2).

Test for https://github.com/basho/riak_core/pull/767 (RIAK-1988)

### CODE REVIEW

**Comment by cmeiklejohn** [comment]
What's the proper way to run this test?

**Comment by cmeiklejohn** [comment]
I have a valid EQC license, but I'm not sure how to get this test to run correctly.

```
================ proxy_overload_recovery failure stack trace =====================
{{assertEqual_failed,
     [{module,proxy_overload_recovery},
      {line,52},
      {expression,"rpc : call ( Node1 , ? MODULE , rtrun , [ TestingTime ] )"},
      {expected,true},
      {value,
          {badrpc,
              {'EXIT',
                  {licence_authorization_failed,eqc,
                      {licence_authorization_failed,eqc_licence,
                          {{wrong_code_file_versions,
                               [eqc_mcerlang,eqc_mce_prop_transform,
                                eqc_mcerlang_statem]},
                           [{eqc_licence,f733768_0,1,
                                [{file,"../src/eqc_licence.erl"},
                                 {line,60}]}]}}}}}}]},
 [{proxy_overload_recovery,'-confirm/0-fun-1-',3,
      [{file,"tests/proxy_overload_recovery.erl"},{line,52}]},
  {proxy_overload_recovery,confirm,0,
      [{file,"tests/proxy_overload_recovery.erl"},{line,52}]},
  {riak_test_runner,return_to_exit,3,
      [{file,"src/riak_test_runner.erl"},{line,159}]}]}
==================================================================================

16:56:27.085 [notice] proxy_overload_recovery Test Run Complete
16:56:27.086 [info] Stopping cover

Test Results:
proxy_overload_recovery-bitcask: fail
---------------------------------------------
1 Tests Failed
0 Tests Passed
That's 0.0% for those keeping score


So, we find ourselves in a tricky situation here. 
You've run a single test, and it has failed.
Would you like to leave Riak running in order to debug?
[Y/n] 
```

**Comment by jonmeredith** [comment]
Yikes.  It magically runs for me and sets the EQC paths.  Until I resolve the rt issues you can just compile inside a running node and just call the run(Seconds) function.

**Comment by jonmeredith** [comment]
@cmeiklejohn it should use the EQC you have available now.

**Comment by cmeiklejohn** [comment]
Doesn't appear to be working.

```
================ proxy_overload_recovery failure stack trace =====================
{{assertEqual_failed,
     [{module,proxy_overload_recovery},
      {line,53},
      {expression,"rpc : call ( Node1 , ? MODULE , rtrun , [ TestingTime ] )"},
      {expected,true},
      {value,
          {badrpc,
              {'EXIT',
                  {licence_authorization_failed,eqc,
                      {licence_authorization_failed,eqc_licence,
                          {{wrong_code_file_versions,
                               [eqc_mcerlang,eqc_mce_prop_transform,
                                eqc_mcerlang_statem]},
                           [{eqc_licence,f733768_0,1,
                                [{file,"../src/eqc_licence.erl"},
                                 {line,60}]}]}}}}}}]},
 [{proxy_overload_recovery,'-confirm/0-fun-1-',3,
      [{file,"tests/proxy_overload_recovery.erl"},{line,53}]},
  {proxy_overload_recovery,confirm,0,
      [{file,"tests/proxy_overload_recovery.erl"},{line,53}]},
  {riak_test_runner,return_to_exit,3,
      [{file,"src/riak_test_runner.erl"},{line,159}]}]}
==================================================================================

19:59:12.131 [notice] proxy_overload_recovery Test Run Complete
19:59:12.139 [info] Stopping cover
```

File: tests/proxy_overload_recovery.erl [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+%%   1) If the vnode message queue has cleared, the vnode proxy should not
+%%   be in overload.
+%%   2) If the proxy has been sent more messages than the overload threshold
+%%      while the vnode is suspended, it should be in overload.
+%%   3) If the proxy has been sent less than or equal to the overload threshold
+%%      it should not be in overload.
+%%
+%% Whenever the vnode is resumed, the test checks the message queues are
+%% cleared and the model is reset.  The main goal is to advance the proxy
+%% into interestign new states.
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Typo 'interestign'

File: tests/proxy_overload_recovery.erl [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+-include_lib("eqc/include/eqc.hrl").
+-include_lib("eqc/include/eqc_statem.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-export([confirm/0]).
+-compile(export_all).
+
+confirm() ->
+    %% disable AAE to prevent vnode start delay opening AAE leveldb
+    Opts = [{riak_kv, [{anti_entropy, {off, []}}]}],
+    Nodes = [Node1] = rt:deploy_nodes(1, Opts),
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Most of the new tests uses the `build_clusters` API, and no longer call `deploy_nodes` directly.

Comment (jonmeredith) [inline tests/proxy_overload_recovery.erl]: What's the best way to deploy a singleton, that's all this test needs.

File: tests/proxy_overload_recovery.erl [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+%%
+%% Run prepare once before any other calls and then never again.
+%%
+precondition_common(#tstate{rt = undefined}, {call, _M, F, _A}) ->
+    F == prepare;
+precondition_common(_, {call, _M, F, _A}) ->
+    F /= prepare.
+
+%% %% Make sure we're still running what we think we're running - uncomment
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Maybe remove this dead code?

File: tests/proxy_overload_recovery.erl [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    catch
+        _:Err ->
+            lager:info("EQC unavailable: ~p\n", [Err]),
+            {error, unavailable}
+    end.
+
+
+%% Add the QuickCheck in the r_t runner to the list of Nodes
+add_eqc_apps(Nodes) ->
+    Apps = [eqc, pulse, pulse_otp],
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Should we commit the `eqc_mcerlang` change?

File: tests/proxy_overload_recovery.erl [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        {'DOWN', Mref, _, _, _} ->
+            ok
+    end.
+
+%%
+%% Check if EQC is available on the local node, then add to remotes if needed
+%% This could be refactored out if many tests need it.
+setup_eqc(Nodes) ->
+    try
+        _ = eqc:reserved_until(), % will throw exception if problems
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Do the builders have eqc?  If not, will this return as a test failure?

File: tests/proxy_overload_recovery.erl [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+
+resume_args(#tstate{rt = RT}) ->
+    [RT].
+
+resume(#rt{ppid = PPid, vpid = VPid}) ->
+    sys:resume(VPid),
+    ok = drain(VPid),
+    %% make sure all messages have been processed and there
+    %% is no pending 'ping' in the vnode before we continue.
+    sys:get_status(PPid),
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: You're calling `get_status` but doing nothing with the response...?

File: tests/proxy_overload_recovery.erl [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+    sys:get_status(VPid).
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Similiar... shouldn't you verify the queue has been drained?

File: tests/proxy_overload_recovery.erl [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+%%
+%% Check the current state of the vnode so we can test properties
+%%
+overloaded_args(#tstate{vnode_running = Running, rt = RT}) ->
+    [Running, RT].
+
+overloaded(Running, #rt{ppid = PPid, vpid = VPid}) ->
+    case Running of
+        true ->
+            ok = drain(PPid), % make sure all proxy msgs procesed/dropped
```
Comment (cmeiklejohn) [inline tests/proxy_overload_recovery.erl]: Spelling `procesed`

**Comment by cmeiklejohn** [comment]
:+1: d43e526

Some minor nitpicking, but a cursory read of the code looks good and is straightforward; model simply randomly sends direct messages vs. proxy messages and verifies that the overload code compensates accordingly.

File: tests/proxy_overload_recovery.erl (hunk 3 above)
Comment (jonmeredith) [inline tests/proxy_overload_recovery.erl]: Which lines do you think are dead? 

The precondition match on rt = undefined makes sure it is always the only task that can be selected first. The second clause makes sure that prepare is never selected again once the runtime parameters are created.

File: tests/proxy_overload_recovery.erl (hunk 4 above)
Comment (jonmeredith) [inline tests/proxy_overload_recovery.erl]: Yes, I need to fix it, but probably make it a better check to see if it exists locally.

File: tests/proxy_overload_recovery.erl (hunk 5 above)
Comment (jonmeredith) [inline tests/proxy_overload_recovery.erl]: The intent was that if eqc was not installed, or the license could not be reserved then we issue a warning saying the test did not actually run and pass it.

Unfortunately there's no way of skipping that I could work out looking at the r_ sources.

File: tests/proxy_overload_recovery.erl (hunk 6 above)
Comment (jonmeredith) [inline tests/proxy_overload_recovery.erl]: Yes, it's a trick to make sure a synchronous call goes through the proxy pid without changing any state, guaranteeing any messages that have been sent will be processed before the next commands.  I'll clarify the comment further.

File: tests/proxy_overload_recovery.erl (hunk 7 above)
Comment (jonmeredith) [inline tests/proxy_overload_recovery.erl]: The thought was - make sure the vnode is drained, which may send pending pong messages back to the proxy. Drain does not guarantee there were not messages pending, so the get_status checks make sure that happens. 

However, I think the order is wrong - this is an implementations specific detail which is why a pulse test would be better, but as the proxy does not respond to pong messages, we should ensure they are sent by the vnode *before* we check there are no pending messages being processed in the proxy.

Nice catch.

**Comment by jonmeredith** [comment]
@cmeiklejohn thanks for the review - fixed things up and pushed.

**Comment by cmeiklejohn** [comment]
:+1: 865a389

Nicely done.

**Comment by cmeiklejohn** [comment]
Does this need to be merged into the mainline as well?  What about GiddyUp?

**Comment by jonmeredith** [comment]
Yes. Will merge it forward and add.

**Pull request merged**